source/extensions/filters/http/oauth2/filter.cc

```diff
@@ -507,6 +540,52 @@ void OAuth2Filter::finishGetAccessTokenFlow() {
   config_->stats().oauth_success_.inc();
 }
 
+void OAuth2Filter::finishUpdateAccessTokenFlow() {
+  ASSERT(config_->useRefreshToken());
+  // At this point we have updated all of the pieces need to authorize a user
+  // We need to actualize keys in the cookie header of the current request related
+  // with authorization. So, the upstream can use updated cookies for itself purpose
+  const CookieNames& cookie_names = config_->cookieNames();
+
+  absl::flat_hash_map<std::string, std::string> cookies =
+      Http::Utility::parseCookies(*request_headers_);
+
+  cookies.insert_or_assign(cookie_names.oauth_hmac_, getEncodedToken());
+  cookies.insert_or_assign(cookie_names.oauth_expires_, new_expires_);
+
+  if (config_->forwardBearerToken()) {
+    cookies.insert_or_assign(cookie_names.bearer_token_, access_token_);
+    if (!id_token_.empty()) {
+      cookies.insert_or_assign(cookie_names.id_token_, id_token_);
+    }
+    if (!refresh_token_.empty()) {
+      cookies.insert_or_assign(cookie_names.refresh_token_, refresh_token_);
+    }
+  }
+
+  std::string new_cookies(absl::StrJoin(cookies, "; ", absl::PairFormatter("=")));
+  request_headers_->addReferenceKey(Http::Headers::get().Cookie, new_cookies);
+  if (config_->forwardBearerToken() && !access_token_.empty()) {
+    setBearerToken(*request_headers_, access_token_);
+  }
+
+  // At this point we add set-cookie headers to the temporary object of the response
+  // The temporary response will populate the response from upstream in the encodeHeaders method
+  // Eventually set-cookie headers will be send to the user agent (browser)
+  response_headers_to_add_ = Http::ResponseHeaderMapImpl::create();
+  addResponseCookies(*response_headers_to_add_, getEncodedToken());
```

Comment:
Why we need create a header map here? I think a simple bool is enough here to tell we should update the response headers?

Reply:
We need to pass a map object to addResponseCookies. So, we create object. addResponseCookies method populate headers to the map

Reply:
I mean why cannot we just call the addResponseCookies in the `encodeHeaders()` and use the accutally response headers as the parameter. Then we needn't create a unnecessary temporary object here. did I miss something?